include/ada/url-inl.h

```diff
@@ -16,166 +16,183 @@
 #endif // ADA_REGULAR_VISUAL_STUDIO
 
 namespace ada {
-  [[nodiscard]] ada_really_inline bool url::includes_credentials() const noexcept {
-    return !username.empty() || !password.empty();
-  }
-  [[nodiscard]] inline bool url::cannot_have_credentials_or_port() const {
-    return !host.has_value() || host.value().empty() || type == ada::scheme::type::FILE;
-  }
-
-  inline std::ostream& operator<<(std::ostream& out, const ada::url& u) {
-    return out << u.to_string();
-  }
-
-  // number of 'leading zeroes'.
-  inline int leading_zeroes(uint32_t input_num) {
+[[nodiscard]] ada_really_inline bool
+url::includes_credentials() const noexcept {
+  return !username.empty() || !password.empty();
+}
+[[nodiscard]] inline bool url::cannot_have_credentials_or_port() const {
+  return !host.has_value() || host.value().empty() ||
+         type == ada::scheme::type::FILE;
+}
+
+inline std::ostream &operator<<(std::ostream &out, const ada::url &u) {
+  return out << u.to_string();
+}
+
+// number of 'leading zeroes'.
+inline int leading_zeroes(uint32_t input_num) {
 #if ADA_REGULAR_VISUAL_STUDIO
-    unsigned long leading_zero(0);
-    unsigned long in(input_num);
-    return _BitScanReverse(&leading_zero, in) ? int(31 - leading_zero) : 32;
+  unsigned long leading_zero(0);
+  unsigned long in(input_num);
+  return _BitScanReverse(&leading_zero, in) ? int(31 - leading_zero) : 32;
 #else
-    return __builtin_clz(input_num);
-#endif// ADA_REGULAR_VISUAL_STUDIO
-  }
-
-  // integer logarithm of x (ceil(log2(x)))
-  inline int int_log2(uint32_t x) {
-    return 31 - leading_zeroes(x | 1);
-  }
-
-  // faster than std::to_string(x).size().
-  inline int fast_digit_count(uint32_t x) {
-    // Compiles to very few instructions. Note that the
-    // table is static and thus effectively a constant.
-    // We leave it inside the function because it is meaningless
-    // outside of it (this comes at no performance cost).
-    const static uint64_t table[] = {
+  return __builtin_clz(input_num);
+#endif // ADA_REGULAR_VISUAL_STUDIO
+}
+
+// integer logarithm of x (ceil(log2(x)))
+inline int int_log2(uint32_t x) { return 31 - leading_zeroes(x | 1); }
```

Comment:
Shouldn't this function be in `helpers`?

Reply:
It is only used in url::get_components() so I think it makes sense to have it defined only in the file where it is used, but let us explain it, and let us grant it internal linkage. We can move it to helpers if it becomes used more broadly.

Generally, we have been abusing system-wise function which may cause the library to have too many entry points. This can harm us later when people come to depend on the existence of a function in the library... we may then be stuck and have to keep supporting.

Reply:
You're right.